### ISSUE TITLE
Categorical kernel can not use stretch()

### ISSUE THREAD

**Issue by patel-zeel** [wesselb/mlkernels#2]
Hi @wesselb,

I was implementing [```CategoricalKernel()```](https://github.com/pytorch/botorch/blob/main/botorch/models/kernels/categorical.py) where I think we can not make use of the ```stretch()``` method. The reason is, ```stretch()``` directly divides inputs by length scales, which is not suitable for ```CategoricalKernel()```. As of now, I could think of the following version. Can you suggest a better idea or am I missing something here?

```
import lab as B
from matrix import Dense

from . import _dispatch
from .. import Kernel

__all__ = ["Categorical"]


class Categorical(Kernel):
    """Categorical kernel based on Hamming distance."""
    
    def __init__(self, scale):
        self.scale = scale

    def _compute(self, dists):
        return B.exp(-1. * dists/self.scale ** 2)

    @property
    def _stationary(self):
        return True

    @_dispatch
    def __eq__(self, other: "Categorical"):
        return True

# After implementing pw_ne and ew_ne in LAB.

@_dispatch
def pairwise(k: Categorical, x: B.Numeric, y: B.Numeric):
    return Dense(k._compute(B.pw_ne(x, y)))


@_dispatch
def elwise(k: Categorical, x: B.Numeric, y: B.Numeric):
    return k._compute(B.ew_ne(x, y))

```

Also, I think this problem could be generalized for all kernels that can not scale the inputs directly (Gibbs kernel in [4.32](http://www.gaussianprocess.org/gpml/chapters/RW4.pdf)).

**Comment by wesselb** [comment]
Yes, I think you're totally right that `stretch` cannot used to implement this kernel. Your proposed solutions looks very reasonable. :)

One comment is that perhaps you don't want to call the parameter `scale` but something else, because, as you point out, the parameter is not really comparable to the usual length scales. (I guess the concept of a length scale is not really defined. One definition which is sometimes used is `scale(k) = ∫ k(r) dr / k(0)`.)

You're also very right in that the Gibbs kernel cannot really be implemented by passing an input-dependent length scale to `stretch`. That's because the construction of the Gibbs kernel is fundamentally slightly different; see the paragraph just below (4.32) in the book. It would be very nice to have a natural way to implement the Gibbs kernel. Do you have any ideas?

**Comment by patel-zeel** [comment]
Thank you for your response, @wesselb. I could not understand ```scale(k) = ∫ k(r) dr / k(0)```. Can you please show an example of that on ```EQ()``` or any other kernel?

I am thinking of the following way that may work for all kernels: What if we use ```Q_ij``` concept given in [Eq. 1 & 2](https://proceedings.neurips.cc/paper/2003/file/326a8c055c0d04f5b06544665d8bb3ea-Paper.pdf)? (maybe given in some other literature as well). I think that will work for any kernel. Though, we may need to check if this makes computation time worse than before.

**Comment by wesselb** [comment]
> Thank you for your response, @wesselb. I could not understand scale(k) = ∫ k(r) dr / k(0). Can you please show an example of that on EQ() or any other kernel?

Definitely! To really define what a length scale is, we need to assign it a definition which works for any kernel, rather than pointing at a particular parameter in a kernel and hand wavily calling that the kernel's length scale. According to the above, for a stationary kernel `k(r)`, the length scale `scale(k)` of the kernel is proposed to be defined by the integral `scale(k) = ∫ k(r) dr / k(0)`. If `k(r) = exp(-r^2 / (2 ell^2))` (EQ kernel), then `scale(k) = sqrt(pi / 2) * ell`; and if `k(r) = exp(-r / ell)` (exponential kernel), then `scale(k) = ell`. Therefore, for the EQ kernel and the exponential kernel, the integral definition agrees with the parameters that we'd usually call the kernel's length scale (up to constants not too far from `1`: `sqrt(pi / 2) \approx 1.25`). The benefit is that we can now take any kernel, which may not have an explicit parameter called the length scale, and compute its length scale with the integral definition.

> I am thinking of the following way that may work for all kernels: What if we use Q_ij concept given in Eq. 1 & 2? (maybe given in some other literature as well). I think that will work for any kernel. Though, we may need to check if this makes computation time worse than before.

Are you proposing to use the concept in (1) and (2) as a general definition of a length scale? This seems to make sense and be similar to an ARD kernel: rather than depending on `||x1 - x2||^2`, depend on `(x1 - x2)^T inverse(A) (x1 - x2)` for some scale matrix `A`. Would the concept in (1) and (2) be different from this?
